### PR TITLE
D3ASIM-3569: Report validation status of myco external matcher recommendations

### PR DESCRIPTION
**Pull request by Hassan754**

### CODE REVIEW

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (Hassan754) [inline src/d3a/models/myco_matcher/myco_external_matcher.py:201]: These were removed because they are duplicates available in the match_recommendations function of the Two Sided market
These are non blockers, they do not stop the matching process, they just omit the recommendation

File: src/d3a/models/market/two_sided.py
Comment (spyrostz) [inline src/d3a/models/market/two_sided.py:359]: If we really want the requirements matching to act as pay as bid, it should be this way IMO:

```suggestion
        if any(not isclose(bid.energy_rate + FLOATING_POINT_TOLERANCE, clearing_rate) for bid in bids):
```

**Review by spyrostz — COMMENTED** [review]
One important clarification pending, but LGTM once this is fixed or discussed.

File: src/d3a/models/market/two_sided.py
Comment (Hassan754) [reply src/d3a/models/market/two_sided.py:359]: In my understanding the match can be valid if the clearing rate is less than or equal to the bid's rate and greater than or equal to the offer's rate

Comment (Hassan754) [reply src/d3a/models/market/two_sided.py:359]: @candrews42

Comment (spyrostz) [reply src/d3a/models/market/two_sided.py:359]: Since this is a generic validator, and not trying to match the 2-sided PAB market, then I agree, you can leave it as is.

**Review by andreabertolini1995 — DISMISSED** [review]
LGTM

**Review by andreabertolini1995 — APPROVED** [review]
LGTM

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (hannesdiedrich) [inline src/d3a/models/myco_matcher/myco_external_matcher.py:238]: Nice! Perfect example for motivation of not using simple `Exception`s, right @fievelk ? This is what you meant in Lisbon, right?

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/myco_external_matcher.py:89]: Because it can also be None. so I marked it as Optional

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/myco_external_matcher.py]: Update, I have removed the broad `except Exception` to not report **any** exception to the client

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/myco_external_matcher.py:234]: As agreed with Spyros and Colin, these blocking exceptions should stop the whole matching process, so no need to further loop over the rest of the recommendations

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (fievelk) [reply src/d3a/models/myco_matcher/myco_external_matcher.py:89]: @hannesdiedrich if you're interested, you can find more information here: https://docs.python.org/3/library/typing.html#typing.Optional

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (fievelk) [reply src/d3a/models/myco_matcher/myco_external_matcher.py]: @hannesdiedrich this can be one of such cases where you want to have a distinction between different types of exceptions, yes!

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (hannesdiedrich) [reply src/d3a/models/myco_matcher/myco_external_matcher.py:89]: Thanks! had a look and am convinced. Can you use the `Dict` from `typing` instead of the build in `dict` though? Thanks!

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (hannesdiedrich) [reply src/d3a/models/myco_matcher/myco_external_matcher.py:234]: Interesting, just for me to understand (not in order to start a long discussion again, which I missed):
If I the myco matcher sends 10 recommendations and the second fails, the first one will be matched but not the other (probably valid) recommendations 3-10 ?

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (hannesdiedrich) [inline src/d3a/models/myco_matcher/myco_external_matcher.py:230]: Can you maybe add an inline comment that explains why you handle the `MycoValidationException` differently than the `InvalidBidOfferPairException`  here?

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/myco_external_matcher.py:234]: I'm validating all recommendations first and then if no blocker exception is raised I start matching the validated records, take a look at the `match_recommendations` method:

```python
response_data.update(
            MycoExternalMatcherValidator.validate_and_report(self, recommendations)) # Here I'm sending all recommendations to validation
        if response_data["status"] != "success": # if this fail, there is a blocking exception raised
            self.myco_ext_conn.publish_json(channel, response_data)
            return
        for recommendation in response_data["recommendations"]:
            try:
                if recommendation["status"] != "success":
                    # skip the failed recommendations (non blocking ones)
                    continue
                market = self.markets_mapping.get(recommendation["market_id"])
                market.match_recommendations([recommendation]) # Here I'm doing the matching

```

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/myco_external_matcher.py:230]: I'm not dealing with `MycoValidationException` other than `InvalidBidOfferPairException`
I'm only dealing with the `BLOCKING_EXCEPTIONS` differently, so if `InvalidBidOfferPairException` is added to the `BLOCKING_EXCEPTIONS` it will be dealt with in the same way

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (hannesdiedrich) [reply src/d3a/models/myco_matcher/myco_external_matcher.py:89]: As far as my github diff tells, this has not been addressed yet.

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (faizan2590) [inline src/d3a/models/myco_matcher/myco_external_matcher.py:217]: `validate_valid_dict `, `validate_market_exists ` & `validate_orders_exist_in_market ` can be changed to private method as called within the same class.

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/myco_external_matcher.py:220]: I see your concern, This can be done indeed, however I see that these two classes are tightly coupled (This validator class will only work with the External Matcher)
So I propose either keeping them both in the same module, or creating a package (I also think that this should be avoided now but can be done if we planned to expand this external matcher functionality to add more classes)
In any case, separating them into 2 modules may cause some circular imports issues (The matcher imports the validator and the validator imports the matcher for type hints, this can be silenced though)
So it's your call, thanks!

File: src/d3a/models/myco_matcher/myco_external_matcher.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/myco_external_matcher.py:116]: Ah okay, I couldn't do that because each recommendation can be targeted to a different market
So you can see the line `market = self.markets_mapping.get(recommendation["market_id"])`, I'm getting the market of each recommendation and posting the recommendation in this market
consider this example:
```python
recommendations = [{"market_id": "market1" ...}, {"market_id": "market2" ...}]
for recommendation in recommendations:
    market = self.markets_mapping.get(recommendation["market_id"]) # will return market1 in first iteration and market2 in the second
    market.match_recommendations([recommendation])
```

**Pull request merged**
